File: lib/languages/nim/recursion.nim
# Simple recursion (factorial)
proc factorial(n: int): int =
  if n <= 1:
    1
  else:
    n * factorial(n - 1)

# Tail recursion (more efficient)
proc factorialTail(n: int, acc = 1): int =
  if n <= 1:
    acc
  else:
    factorialTail(n - 1, n * acc)

# Recursive sequence processing
proc sum(numbers: seq[int]): int =
  if numbers.len == 0:
    0
  else:
    numbers[0] + sum(numbers[1..^1])

# Tail recursive sequence processing
proc sumTail(numbers: seq[int], acc = 0): int =
  if numbers.len == 0:
    acc
  else:
    sumTail(numbers[1..^1], acc + numbers[0])

# Recursive data structure (binary tree)
type
  Tree = ref object
    value: int
    left, right: Tree

# Recursive tree traversal (in-order)
proc inOrder(tree: Tree): seq[int] =
  if tree == nil:
    @[]
  else:
    result = inOrder(tree.left)
    result.add(tree.value)
    result = result & inOrder(tree.right)

# Recursive binary search
proc binarySearch(arr: seq[int], target: int, left = 0, right = -1): int =
  var r = if right == -1: arr.len - 1 else: right
  
  if left > r:
    return -1  # Not found
  
  let mid = (left + r) div 2
  
  if arr[mid] == target:
    return mid  # Found
  elif arr[mid] > target:
    return binarySearch(arr, target, left, mid - 1)  # Search left half
  else:
    return binarySearch(arr, target, mid + 1, r)  # Search right half

# Recursive string reversal
proc reverse(s: string): string =
  if s.len <= 1:
    s
  else:
    reverse(s[1..^1]) & s[0..0]

# Memoization for optimization
import tables

proc fibMemo(n: int, memo: var Table[int, int]): int =
  if n in memo:
    return memo[n]
  
  if n <= 1:
    result = n
  else:
    result = fibMemo(n-1, memo) + fibMemo(n-2, memo)
  
  memo[n] = result

proc fibonacci(n: int): int =
  var memo = initTable[int, int]()
  fibMemo(n, memo)

when isMainModule:
  echo "factorial(5): ", factorial(5)
  echo "factorialTail(5): ", factorialTail(5)
  
  let numbers = @[1, 2, 3, 4, 5]
  echo "sum(numbers): ", sum(numbers)
  echo "sumTail(numbers): ", sumTail(numbers)
  
  let tree = Tree(
    value: 4,
    left: Tree(
      value: 2,
      left: Tree(value: 1),
      right: Tree(value: 3)
    ),
    right: Tree(
      value: 6,
      left: Tree(value: 5),
      right: Tree(value: 7)
    )
  )
  
  echo "inOrder(tree): ", inOrder(tree)
  
  let sorted = @[1, 3, 5, 7, 9, 11, 13, 15]
  echo "binarySearch(sorted, 7): ", binarySearch(sorted, 7)
  echo "binarySearch(sorted, 6): ", binarySearch(sorted, 6)
  
  echo "reverse(\"hello\"): ", reverse("hello")
  
  echo "fibonacci(10): ", fibonacci(10)

File: lib/languages/kotlin/recursion.kt
// Simple recursion (factorial)
fun factorial(n: Int): Int {
   // Base case
   if (n == 0 || n == 1) {
       return 1
   }
   // Recursive case
   return n * factorial(n - 1)
}

// Tail recursion (more efficient)
tailrec fun factorialTail(n: Int, acc: Int = 1): Int {
   // Base case
   if (n == 0 || n == 1) {
       return acc
   }
   // Tail-recursive case
   return factorialTail(n - 1, n * acc)
}

// Recursive Fibonacci
fun fibonacci(n: Int): Int {
   // Base cases
   if (n <= 1) {
       return n
   }
   // Recursive case
   return fibonacci(n - 1) + fibonacci(n - 2)
}

// Tail-recursive Fibonacci
tailrec fun fibonacciTail(n: Int, a: Int = 0, b: Int = 1): Int {
   if (n == 0) return a
   if (n == 1) return b
   return fibonacciTail(n - 1, b, a + b)
}

// Recursive list processing
fun sum(list: List<Int>): Int {
   if (list.isEmpty()) {
       return 0
   }
   return list.first() + sum(list.drop(1))
}

// Tail-recursive list processing
tailrec fun sumTail(list: List<Int>, acc: Int = 0): Int {
   if (list.isEmpty()) {
       return acc
   }
   return sumTail(list.drop(1), acc + list.first())
}

// Recursive binary search
fun binarySearch(array: IntArray, target: Int, left: Int = 0, right: Int = array.size - 1): Int {
   if (left > right) {
       return -1  // Not found
   }
   
   val mid = left + (right - left) / 2
   
   return when {
       array[mid] == target -> mid  // Found
       array[mid] > target -> binarySearch(array, target, left, mid - 1)  // Search left half
       else -> binarySearch(array, target, mid + 1, right)  // Search right half
   }
}

// Recursive tree traversal
class TreeNode(val value: Int, val left: TreeNode? = null, val right: TreeNode? = null)

fun inOrderTraversal(node: TreeNode?, result: MutableList<Int> = mutableListOf()): List<Int> {
   if (node == null) {
       return result
   }
   
   inOrderTraversal(node.left, result)
   result.add(node.value)
   inOrderTraversal(node.right, result)
   
   return result
}

